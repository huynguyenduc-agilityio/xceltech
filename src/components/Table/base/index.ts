export * from './TableBase';
export * from './TableHeader';
export * from './TableBody';
export * from './TableRow';
export * from './TableCell';
