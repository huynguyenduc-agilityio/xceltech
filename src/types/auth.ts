export interface ILoginUser {
  email: string;
  password: string;
}

export interface IAuthUser {
  refresh: string;
  access: string;
  user: {
    id: string;
    email: string;
  };
}

export interface IRegisterUser extends ILoginUser {
  firstName: string;
  lastName: string;
  phone: string;
  confirmPassword: string;
  isAgreeTerms?: boolean;
  isReceiveNewsletters?: boolean;
}

export interface IActiveUser {
  uidb64: string | null;
  token: string | null;
}
