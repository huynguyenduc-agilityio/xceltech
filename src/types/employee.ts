export interface IEmployeePersonalInfo {
  name: string;
  avatar?: string;
  department: string;
  jobTitle: string;
  jobCategory: string;
}
