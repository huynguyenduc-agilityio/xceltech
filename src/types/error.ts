export type ErrorType = {
  detail: string;
};
