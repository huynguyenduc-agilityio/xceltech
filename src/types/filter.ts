export interface FilterCriteria {
  [key: string]: string | number | boolean | string[];
}
