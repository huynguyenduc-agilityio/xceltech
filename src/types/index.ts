export * from './status';
export * from './employee';
export * from './table';
export * from './auth';
export * from './form';
export * from './tab';
