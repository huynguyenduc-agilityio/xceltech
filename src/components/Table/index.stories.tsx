import type { Meta, StoryObj } from '@storybook/react';

// Components
import Table from '.';

const meta = {
  title: 'Components/Table',
  component: Table,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  render: (args) => (
    <div className="flex justify-center h-screen">
      <div className="w-[1000px]">
        <Table {...args} />
      </div>
    </div>
  ),
} satisfies Meta<typeof Table>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    columns: [
      {
        title: 'Name(s)',
        key: 'name',
      },
      {
        title: 'Duration(s)',
        key: 'duration',
      },
      {
        title: 'Start Date',
        key: 'startDate',
      },
      {
        title: 'End Date',
        key: 'endDate',
      },
      {
        title: 'Type',
        key: 'type',
      },
      {
        title: 'Reason(s)',
        key: 'reason',
      },
    ],
    dataSource: [
      {
        id: '1',
        name: 'John Caster',
        duration: 5,
        startDate: '22/04/2022',
        endDate: '28/04/2022',
        type: 'Sick',
        reason: 'Personal',
      },
      {
        id: '2',
        name: 'aman bey',
        duration: 7,
        startDate: '22/04/2022',
        endDate: '28/04/2022',
        type: 'Exam',
        reason: 'Examination',
      },
      {
        id: '3',
        name: 'feven Tesfaye',
        duration: 120,
        startDate: '22/04/2022',
        endDate: '28/04/2022',
        type: 'Maternity',
        reason: 'Child Care',
      },
    ],
  },
};
