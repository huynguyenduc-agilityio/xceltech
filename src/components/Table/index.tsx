import { useMemo } from 'react';

// Types
import { TDataSource, THeaderTable } from '@/types';

// Utils
import { cn, processTableData } from '@/utils';

// Components
import { TableBase, TableHeader, TableBody, TableRow, TableCell } from './base';
import FilterMenu from '../FilterMenu';

type TableProps = {
  title?: string;
  isLoading?: boolean;
  columns?: THeaderTable[];
  dataSource?: TDataSource[];
};

const Table = ({
  title = 'Leave History',
  isLoading,
  columns = [],
  dataSource = [],
}: TableProps) => {
  const processedData = useMemo(
    () => processTableData(columns, dataSource),
    [columns, dataSource],
  );

  return (
    <div className="w-full overflow-hidden">
      <div className="flex justify-between px-5 py-6">
        <h2 className="text-xl font-bold">{title}</h2>
        {/* TODO: Apply filter */}
        <FilterMenu options={['type']} onApply={() => {}} />
        {/* TODO: Add Button Export  */}
      </div>

      <TableBase>
        <TableHeader>
          <TableRow className="bg-blue-light">
            {columns.map(({ key, title }) => (
              <TableCell key={key} className="py-6 text-center">
                {title}
              </TableCell>
            ))}
          </TableRow>
        </TableHeader>
        <TableBody>
          {processedData.length ? (
            processedData.map((row, index) => (
              <TableRow
                key={row.id}
                className={cn(
                  'px-5',
                  index % 2 === 0 ? 'bg-white' : 'bg-blue-light',
                )}
              >
                {row.cells.map((cell) => (
                  <TableCell key={cell.key} className="py-6 text-center">
                    {cell.content}
                  </TableCell>
                ))}
              </TableRow>
            ))
          ) : (
            <TableRow>
              <TableCell colSpan={columns.length} className="text-center">
                {isLoading ? 'Loading...' : 'Empty data'}
              </TableCell>
            </TableRow>
          )}
        </TableBody>
      </TableBase>
    </div>
  );
};

export default Table;
