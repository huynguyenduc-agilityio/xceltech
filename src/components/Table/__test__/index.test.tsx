import { ReactNode } from 'react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';

// Types
import { TDataSource } from '@/types';

// Components
import Table from '..';

describe('Table Component', () => {
  const queryClient = new QueryClient();

  const Wrapper = ({ children }: { children: ReactNode }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  const dataSource: TDataSource[] = [
    {
      id: '1',
      name: 'John Caster',
      duration: 5,
      startDate: '22/04/2022',
      endDate: '28/04/2022',
      type: 'Sick',
      reason: 'Personal',
    },
    {
      id: '2',
      name: 'aman bey',
      duration: 7,
      startDate: '22/04/2022',
      endDate: '28/04/2022',
      type: 'Exam',
      reason: 'Examination',
    },
    {
      id: '3',
      name: 'feven Tesfaye',
      duration: 120,
      startDate: '22/04/2022',
      endDate: '28/04/2022',
      type: 'Maternity',
      reason: 'Child Care',
    },
  ];

  const renderSetup = () =>
    render(
      <Wrapper>
        <Table
          columns={[
            {
              title: 'Name(s)',
              key: 'name',
            },
            {
              title: 'Duration(s)',
              key: 'duration',
            },
            {
              title: 'Start Date',
              key: 'startDate',
            },
            {
              title: 'End Date',
              key: 'endDate',
            },
            {
              title: 'Type',
              key: 'type',
            },
            {
              title: 'Reason(s)',
              key: 'reason',
            },
          ]}
          dataSource={dataSource}
        />
      </Wrapper>,
    );

  it('should match snapshot', () => {
    const { container } = renderSetup();

    expect(container).toMatchSnapshot();
  });
});
