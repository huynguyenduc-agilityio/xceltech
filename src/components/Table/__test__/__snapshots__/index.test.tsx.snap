// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component should match snapshot 1`] = `
<div>
  <div
    class="w-full overflow-hidden"
  >
    <div
      class="flex justify-between px-5 py-6"
    >
      <h2
        class="text-xl font-bold"
      >
        Leave History
      </h2>
      <button
        aria-controls="radix-:r0:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class=""
        data-state="closed"
        type="button"
      >
        <button
          class="font-productsans inline-flex items-center justify-center whitespace-nowrap text-lg font-bold ring-offset-background transition-all duration-300 ease-in-out outline-none focus-visible:outline-none disabled:opacity-50 disabled:cursor-not-allowed hover:bg-accent hover:text-accent-foreground h-10 w-10 rounded-full p-2"
        >
          <svg
            fill="none"
            height="25"
            viewBox="0 0 31 25"
            width="31"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#a)"
            >
              <path
                d="M29.545 0H1.455C.165 0-.487 1.263.427 2l11.198 9.032v10.062c0 .382.231.74.62.96l4.844 2.733c.955.54 2.286-.007 2.286-.96V11.032l11.198-9.031C31.486 1.264 30.839 0 29.545 0"
                fill="#1A1A1A"
              />
            </g>
            <defs>
              <clippath
                id="a"
              >
                <path
                  d="M0 0h31v25H0z"
                  fill="#fff"
                />
              </clippath>
            </defs>
          </svg>
        </button>
      </button>
    </div>
    <div
      class="relative w-full overflow-auto"
    >
      <table
        class="w-full caption-bottom text-sm"
      >
        <thead
          class="text-lg font-bold"
        >
          <tr
            class="transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted bg-blue-light"
          >
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              Name(s)
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              Duration(s)
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              Start Date
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              End Date
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              Type
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              Reason(s)
            </td>
          </tr>
        </thead>
        <tbody
          class="[&_tr:last-child]:border-0 text-lg"
        >
          <tr
            class="transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted px-5 bg-white"
          >
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              John Caster
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              5
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              22/04/2022
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              28/04/2022
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              Sick
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              Personal
            </td>
          </tr>
          <tr
            class="transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted px-5 bg-blue-light"
          >
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              aman bey
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              7
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              22/04/2022
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              28/04/2022
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              Exam
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              Examination
            </td>
          </tr>
          <tr
            class="transition-colors hover:bg-muted/50 data-[state=selected]:bg-muted px-5 bg-white"
          >
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              feven Tesfaye
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              120
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              22/04/2022
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              28/04/2022
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              Maternity
            </td>
            <td
              class="p-2 align-middle [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] py-6 text-center"
            >
              Child Care
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
